### PR TITLE
Setup Heroku deploys and DB

### PR DESCRIPTION
**Pull request by rarmatei**
#5 #2

### CODE REVIEW

**Review by kabaros — APPROVED** [review]

File: README.md [hunk 1]
```diff
@@ -4,28 +4,34 @@
...
+
+- run `npm run install` in root (this will `npm install` on both client and server)
+- run `npm run dev` in root - You might see some error and the API calls to the database won't work though! Let's fix that!
+- `cd` into `server` and `npm run initial-db-setup` (this will create and populate your DB)
+- try out the *Development Workflow* below
+
+Your team will now want their own DB name, tables and data. So one team member needs to:
+- open `server/__db/initial-setup.sql` and replace `cyf_test` with your own DB name
+- open `server/__db/db-config.js` and replace `cyf_test` with your own DB name
+- open `recreate-db.sql` and your own DB schema and give it some test data
```
Comment (djgrant) [inline README.md:17]: "and _add_ your own"?

File: README.md [hunk 2]
```diff
@@ -4,28 +4,34 @@
...
+- commit and push above changes
+
+Once the other team members go the above you'll need to:
+- `cd` into `server` and `npm run initial-db-setup` (this will create and populate your new team's DB with the data your colleague added)
+
```
Comment (djgrant) [inline README.md:22]:
```
- remove the "Initial setup" instructions from the README
```

File: TECHNICAL.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+
+- The communication between server and client works when developing because of adding a proxy value on `package.json` client-side, which allows the calls to API to be made without specifying the hostname. On Heroku, there is a line on the server `app.js` that makes it possible to use `express` to serve the React app serving it from the `build` folder (heroku runs `npm run build` first).
+- The app is published by Heroku CI workflow on: [https://project-skeleton.herokuapp.com](https://project-skeleton.herokuapp.com)
+- The PRs should also be published automatically
```
Comment (djgrant) [inline TECHNICAL.md:4]:
```
- PRs should be published to [Heroku Review Apps](https://devcenter.heroku.com/articles/github-integration-review-apps)
```

File: TECHNICAL.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+
+- The communication between server and client works when developing because of adding a proxy value on `package.json` client-side, which allows the calls to API to be made without specifying the hostname. On Heroku, there is a line on the server `app.js` that makes it possible to use `express` to serve the React app serving it from the `build` folder (heroku runs `npm run build` first).
```
Comment (djgrant) [inline TECHNICAL.md:2]: New bullet for "On Heroku..."

Comment (djgrant) [reply TECHNICAL.md:2]: Something like "...without specifying the _server_ hostname e.g. x.com -> y.com" would be help to build a mental model here.

File: client/src/api/status.js [hunk 5]
```diff
@@ -1,3 +1,11 @@
 export const getStatus = () => {
     return fetch('/api/status').then(res => res.text())
+}
+
+export const getUsers = () => {
+    return fetch('/api/users');
```
Comment (djgrant) [inline client/src/api/status.js:6]: Add `.then(res => res.text())` to these new fns?

File: client/src/containers/Status.js [hunk 6]
```diff
@@ -1,15 +1,25 @@
...
+    hotels: [],
+    users: []
   }
 
   componentDidMount() {
     getStatus().then(response => {
       this.setState({status: response})
     })
+    getHotels().then(response => {
+      this.setState({hotels: response.data});
```
Comment (djgrant) [inline client/src/containers/Status.js:16]: Looks like we need to set up prettier to run here

File: client/src/containers/Status.js [hunk 7]
```diff
@@ -1,15 +1,25 @@
 import React from 'react';
-import { getStatus } from '../api/status';
+import { getStatus, getUsers, getHotels } from '../api/status';
 
 class Status extends React.Component {
   state = {
-    status: null
+    status: null,
+    hotels: [],
```
Comment (djgrant) [inline client/src/containers/Status.js:7]: I think adding these extra endpoints might actually be more confusing for students. Even as an experienced developer I often struggle with starter apps that have too much example code.

Comment (djgrant) [reply client/src/containers/Status.js:7]: I.e. code I need to refactor or delete

File: server/__db/recreate-db.sql [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+CREATE TABLE bookings (
+  id            SERIAL PRIMARY KEY,
+  customer_id   INT REFERENCES customers(id),
+  hotel_id      INT REFERENCES hotels(id),
+  checkin_date  DATE NOT NULL,
+  nights        INT NOT NULL
+);
+
+INSERT INTO customers (name, email, address, city, postcode, country) VALUES ('John Smith','j.smith@johnsmith.org','11 New Road','Liverpool','L10 2AB','UK');
```
Comment (djgrant) [inline server/__db/recreate-db.sql:31]: Can we move this into a separate `populate-db.sql` script?

File: server/__db/db-config.js [hunk 9]
```diff
@@ -0,0 +1,12 @@
+module.exports = {
```
Comment (djgrant) [inline server/__db/db-config.js:1]: Would be worthwhile returning a config based on the current environment so you don't have to do it at the point you import the config.

Something like:
```js
const config = {
  production: require('./production'),
  development: require('./development'),
};

const env = process.env.NODE_ENV;

// could add a check here to make sure a config exists for the env if desired
module.exports = config[env];
```

File: server/__db/initial-setup.sql [hunk 10]
```diff
@@ -0,0 +1,2 @@
+drop database if exists cyf_test;
```
Comment (djgrant) [inline server/__db/initial-setup.sql:1]: It'd be really helpful to indicate in the file name that this is just a local task and not something that runs on heroku. Maybe `set-up-local-db.sql`.

File: server/package.json [hunk 11]
```diff
@@ -7,9 +7,10 @@
     "start": "node app.js",
     "test": "echo \"Error: no test specified\" && exit 0",
     "lint": "prettier-standard './**/*.js'",
-    "dev": "nodemon app.js",
-    "create-db": "echo \"ToDo: create DB\"",
-    "ci-create-db": "echo \"ToDo: create DB\""
+    "dev": "NODE_ENV=development nodemon app.js",
+    "recreate-db:prod": "NODE_ENV=production node ./__db/seed-db.js",
+    "recreate-db": "node ./__db/seed-db.js",
```
Comment (djgrant) [inline server/package.json:12]: How about replacing with: 

```
"recreate-db:local": "psql -U postgres -f __db/set-up-local-db.sql && node ./__db/seed-db.js"`
```
and removing `initial-db-setup`, particulary to save confusion about whether than needs to run on heroku or not.

File: server/helpers/db.js [hunk 12]
```diff
@@ -1,7 +1,21 @@
+const { Pool } = require("pg");
```
Comment (djgrant) [inline server/helpers/db.js]: Alternatively as a shell script, I _think_ you can do this:

```sh
NODE_ENV == production
if [[ -n "$DATABASE_CONNECTION ]]; then
  psql $DATABASE_CONNECTION -f ../__db/db-config
else
  psql -U postgres -f ../__db/db-config
fi




```

File: server/__db/db-config.js (hunk 9 above)
Comment (djgrant) [reply server/__db/db-config.js:1]: I'd also suggest 
a) moving this into the src folder as it's config for the app not the db
b) making it a generic `config.js` file

File: server/__db/seed-db.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+const { Pool } = require("pg");
```
Comment (djgrant) [inline server/__db/seed-db.js:1]: Alternatively as a shell script, I _think_ you can do this:

```sh
if [[ -n "$DATABASE_CONNECTION ]]; then
  psql $DATABASE_CONNECTION -f ../__db/db-config
else
  psql -U postgres -f ../__db/db-config
fi
```

File: server/helpers/db.js [hunk 14]
```diff
@@ -1,7 +1,21 @@
...
+const config = require("../__db/db-config")[process.env.NODE_ENV || "development"];
+
+const pool = new Pool(config);
+
+const getHotels = () => {
```
Comment (djgrant) [inline server/helpers/db.js:6]: I think a db helper should just expose methods to transact with the DB. `getHotels` feels more like it belongs in a a service specifically for interacting with hotel entities.

**Review by djgrant — CHANGES_REQUESTED** [review]

**Comment by djgrant** [comment]
Slightly stylistic preference: can we have a `database` or `db` folder instead of `__db`?

File: server/__db/seed-db.js (hunk 13 above)
Comment (kabaros) [reply server/__db/seed-db.js:1]: wouldn't shell scripts be a problem for the few on windows?

Comment (djgrant) [reply server/__db/seed-db.js:1]: Depends if they've got bash enabled

**Pull request merged**